File: python/chuong2/det.py
import numpy as np


# Hàm tính định thức của ma trận
def tinhDinhThuc(A):
    return np.linalg.det(A)


def transpose(matrix: list[list[float]]) -> list[list[float]]:
    return [[row[col_index] for row in matrix] for col_index in range(len(matrix[0]))]


# Ví dụ sử dụng
A = np.array([[1, 2, 3], [0, 1, 4], [5, 6, 0]])

dinhThuc = tinhDinhThuc(A)
print("Định thức của ma trận A là:", dinhThuc)

transposed = transpose(A)
print("Ma trận chuyển vị là:")
for row in transposed:
    print(row)

File: Chuong2/det.ts
function det(matrix: number[][]): number {
    const n = matrix.length;
    if (n === 1) return matrix[0][0];
    if (n === 2)
        return matrix[0][0] * matrix[1][1] - matrix[0][1] * matrix[1][0];

    let result = 0;
    for (let i = 0; i < n; i++) {
        const subMatrix = matrix
            .slice(1)
            .map((row) => row.filter((_, colIndex) => colIndex !== i));
        result += matrix[0][i] * det(subMatrix) * (i % 2 === 0 ? 1 : -1);
    }
    return result;
}

export default det;
console.log(
    det([
        [1, 2, 3],
        [0, 1, 4],
        [5, 6, 0],
    ])
);
